src/puffin/src/file_format/reader/footer.rs

```diff
@@ -0,0 +1,261 @@
+
+    /// Flags from the footer, set when the `Flags` field is parsed.
+    flags: Flags,
+
+    /// Size of the footer's payload, set when the `PayloadSize` is parsed.
+    payload_size: u64,
+
+    /// Raw payload data to be deserialized into file metadata,
+    /// set when the `Payload` is parsed.
+    payload: Vec<u8>,
+}
+
+/// Represents a read operation that needs to be performed, including the
+/// offset from the start of the file and the number of bytes to read.
+struct BytesToRead {
+    offset: u64,
+    size: u64,
+}
+
+impl StageParser {
+    fn new(file_size: u64) -> Self {
+        Self {
+            stage: ParseStage::FootMagic,
+            file_size,
+            payload_size: 0,
+            flags: Flags::empty(),
+            payload: vec![],
+        }
+    }
+
+    /// Determines the next segment of bytes to read based on the current parsing stage.
+    /// This method returns information like the offset and size of the next read,
+    /// or None if parsing is complete.
+    fn next_to_read(&self) -> Option<BytesToRead> {
+        if self.stage == ParseStage::Done {
+            return None;
+        }
+
+        let btr = match self.stage {
+            ParseStage::FootMagic => BytesToRead {
+                offset: self.foot_magic_offset(),
+                size: MAGIC.len() as _,
+            },
+            ParseStage::Flags => BytesToRead {
+                offset: self.flags_offset(),
+                size: 4,
+            },
+            ParseStage::PayloadSize => BytesToRead {
+                offset: self.payload_size_offset(),
+                size: 4,
+            },
+            ParseStage::Payload => BytesToRead {
+                offset: self.payload_offset(),
+                size: self.payload_size as _,
+            },
+            ParseStage::HeadMagic => BytesToRead {
+                offset: self.head_magic_offset(),
+                size: MAGIC.len() as _,
+            },
+            _ => unreachable!(),
+        };
+
+        Some(btr)
+    }
+
+    /// Processes the bytes that have been read according to the current parsing stage
+    /// and advances the parsing stage. It ensures the correct sequence of bytes is
+    /// encountered and stores the necessary information in the `StageParser`.
+    fn consume_bytes(&mut self, bytes: Vec<u8>) -> Result<()> {
+        match self.stage {
+            ParseStage::FootMagic => {
+                ensure!(bytes == MAGIC, MagicNotMatchedSnafu);
+                self.stage = ParseStage::Flags;
+            }
+            ParseStage::Flags => {
+                self.flags = Self::parse_flags(&bytes)?;
+                self.stage = ParseStage::PayloadSize;
+            }
+            ParseStage::PayloadSize => {
+                self.payload_size = Self::parse_payload_size(&bytes)?;
+                self.stage = ParseStage::Payload;
+            }
+            ParseStage::Payload => {
+                self.payload = self.parse_payload(bytes)?;
```

Comment:
Can we deserialize the payload inside `parse_payload()`? Then we don't need to take and keep the payload buffer in the reader.

e.g. We can just hold the metadata in the reader.
```
self.metadata = Some(self.parse_payload(&bytes)?);
```